packages/cli-e2e/utils/terminal.ts

```diff
@@ -0,0 +1,176 @@
+import {SpawnOptionsWithoutStdio} from 'child_process';
+import {ChildProcessWithoutNullStreams} from 'child_process';
+import {Readable} from 'stream';
+import stripAnsi from 'strip-ansi';
+import {ProcessManager} from './processManager';
+
+type ITerminalSpawnOptions = [
+  string,
+  ReadonlyArray<string>,
+  SpawnOptionsWithoutStdio
+];
+
+export class Terminal {
+  private process: ChildProcessWithoutNullStreams;
+  public orchestrator: Orchestrator;
+
+  constructor(
+    private processManager: ProcessManager = new ProcessManager(),
+    private spawnOptions: ITerminalSpawnOptions
+  ) {
+    this.process = processManager.spawn(...spawnOptions);
+    this.orchestrator = new Orchestrator(this.process);
+  }
+
+  public when(conditionOrCallback: Condition): When {
+    return this.orchestrator.getNewHandle().when(conditionOrCallback);
+  }
+}
+
+class Action {
+  public parallelAction?: Action;
+  public sequentialAction?: Action;
+  public when?: When;
+  public on?: On;
+
+  constructor(public process: ChildProcessWithoutNullStreams) {}
+}
+
+type DoCallback = (
```

Comment:
personal pet peeves, but I generally prefer to see all the `types` / `interface` at the beginning of the file, grouped together.

Reply:
Yeah, for now I wanted to shoot the idea out there about the API etc, the code clearly need to be refined, I'm on it ;)